### ISSUE TITLE
swarm service update pre-hook

### ISSUE THREAD

**Issue by port22** [moby/moby#35494]
Hi all,
I am maintaining nginx configs via git with a post-receive-hook informing swarm to do a service update.
Everytime I git push a new configuration, swarm successfully does this update sequentially one after another. so far so good.
Now, everytime a user is connected to an nginx container, the connection is dropped which is fatal if that happens in an order process, but in the nature of a container being restarted, that's logically.

So my solution to this would be either to not update the service itself and build someting complicated so that the nginx within container pulls the new config and reloads. Well, not my favourite.

There's a consul cluster running and I'd like to use the service feature. With that, I am hoping that I can consul maint -disable the specific node on which the next nginx container is running before swarm service update. 

Is there anything planned to supply hooks at this point?

**Comment by port22** [comment]
My first idea was to write a wrapper which first does consul maint -disable, but I have no idea how to do that for every container separately ...

**Comment by port22** [comment]
yes, a little ... except the fact that the nginx itself is not contacted via ingress (there's HAProxy L4 before nginx with comsul_service_ip as backend). Even if nginx will be stopped "gracefully", there's a chance that HAProxy has still some open connections to nginx. So I wanted to disable the service, sleep some seconds  until HAProxy outruns the open connections, and then do the nginx service update in another per-container-delay of some seconds while re-enabling the updated nginx via consul again

**Comment by red246** [comment]
Is this issue still a problem?
